### DISCUSSION TITLE
How is this different from Napier?

### DISCUSSION THREAD

**Discussion by kawpii** [touchlab/Kermit#217] [Q&A] [answered]
https://github.com/AAkira/Napier

**Comment by kpgalligan** [comment, marked as answer]
Technically speaking or why didn't we use Napier?

Some thoughts:

- Kermit supports as wide of a target set as we reasonably can, which includes various flavors of linux and windows.
- I am *very* critical of native implementations of anything due to my experience around concurrency, etc. Earlier versions of Napier used `@ThreadLocal` for logging config, which meant that logging from other threads did not have access to that config and would log to the default (or nothing). When I first looked at Napier I saw that, and I didn't really dig deeper. That issue has been fixed, but well after we published (earlier versions of) Kermit.
- On a similar point, we were very concerned with performance, and currently Napier appears to route all access through an atomic reference (https://github.com/AAkira/Napier/blob/master/napier/src/commonMain/kotlin/io/github/aakira/napier/Napier.kt#L16). It's not the end of the world, but we had different design goals. Using global access from Kermit does something similar because any library would need to (although it's technically volatile on the JVM), but you can also use static local config which avoids all of that entirely, and there are very few use cases that need mutable logging config in a production app.
- For calling from non-Kotlin platforms we wanted to make the API easy to use. So, say for Swift, you need to provide all default params to Kotlin. We kept the API minimal for that reason, but it's a tradeoff. Looking at Napier code, it looks like you'd need to supply your own Swift extensions (https://github.com/AAkira/Napier/blob/master/ios/Napier/Napier.swift). Better/worse? I think until there is better control over the generated interface from the compiler everybody will need to make some tradeoff decisions, and we made a different one.
- [Crash reporting integrations, with symbolication](https://touchlab.co/kermit-and-crashlytics/)

That's a quick overview. I would need a lot more time to look through and find where other parts of the design differ, but are they both logging libraries that try to do roughly the same thing? Yes!

Historically speaking, when we made Kermit we weren't excited about the other options at the time. Napier has matured in that time, but we have some different API and technical designs internally. We also liked [some of the ideas](https://github.com/touchlab/Kermit/issues/183).

Either can handle shared logging for the vast majority of cases.
